Augment AST before generating module

nbhuiyan opened the pull request:
Some of the AST nodes now have a functionEntry member that
provides a way to lookup symbols defined in parent scopes.
FunctionContext is used for context tracking as well
as looking up symbols beyond local context through the use
of a stack. Function indexing starts from 1, as index 0 is
reserved for the top-level body, which is now treated as a
function.

Also-by: Robert Young <rwy0717@gmail.com>
Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Augment AST before generating module

Some of the AST nodes now have a functionEntry member that
provides a way to lookup symbols defined in parent scopes.
FunctionContext is used for context tracking as well
as looking up symbols beyond local context through the use
of a stack. Function indexing starts from 1, as index 0 is
reserved for the top-level body, which is now treated as a
function.

Also-by: Robert Young <rwy0717@gmail.com>
Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

rwy7 commented on js_compiler/compile.js:
```diff
@@ -330,16 +426,97 @@ function Module() {
+		this.augmentAST(syntax);
+		this.functionContext.enterFunction(syntax.functionEntry);
 		this.handleBody(func, syntax.body);
+		func.instructions.push(new Instruction("END_SECTION", 0)); //end of top level function
```
Can we insert a `(int-push-const 0) (return)` before the `end-section` marker?

nbhuiyan replied:
sure. that would make the `<body>` non-empty, and I strongly believe jitting that would not be a problem.

rwy7 commented on js_compiler/compile.js:
```diff
@@ -307,16 +406,13 @@ function Module() {
+			outputString(out, func.name);
+			outputUInt32(out, func.index);
+			func.output(out);
+		}
```
Was this just a simplification of the old code? Is the behaviour the same? (It looks like it is)

nbhuiyan replied:
Yup. Basically, manual inlining.

rwy7 commented on js_compiler/compile.js:
```diff
@@ -125,6 +125,150 @@ var SymbolTable = function () {
+		if (id !== undefined) {
+			return { type: "function", id: id };
+		}
+		//console.warn("Did not find symbol: " + symbol + " in: " + JSON.stringify(this, null, 2));
```
Stale comment should be deleted

rwy7 commented on b9/src/ExecutionContext.cpp:
```diff
@@ -189,7 +189,8 @@ StackElement ExecutionContext::interpret(const std::size_t functionIndex) {
   }
-  throw std::runtime_error("Reached end of function");
+  //throw std::runtime_error("Reached end of function");
+  return StackElement(0);
```
I think we should preserve the old behaviour: crash if we try to execute an end-section. Eventually, I suspect we'll be eliminating end-section entirely.

What we should do instead is, output an implicit
```lisp
(int-push-const 0)
(return)
```
at the end of the script's body.

rwy7 replied:
Once we start emitting push-0/return implicitly in the body, then it will be impossible for the body function to be empty. I bet this will fix our crash when JIT compiling the body.

rwy7 commented on b9/src/VirtualMachine.cpp:
```diff
@@ -99,12 +99,13 @@ std::size_t VirtualMachine::getFunctionCount() {
   assert(cfg_.jit);
-  auto functionIndex = 0;
+  compiledFunctions_.push_back(nullptr); //TODO: <body> function right now is "compiled" as null
+  auto functionIndex = 1; //0 index for <body>
```
I would rather fix the compilation issue in this PR, than skip over compiling <body>.

nbhuiyan replied:
I agree.

nbhuiyan committed:
Fix issue with JIT-ing top level body

JIT-ing failure was caused by the empty "body" function.
This has been fixed by making it return 0.

Also-by: Robert Young <rwy0717@gmail.com>
Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan commented:
@rwy0717 You're right! Inserting the instructions so that `<body>` returns 0 fixes the issue with jitting it. Requesting another review.

rwy7 commented:
Thanks! Great work!

A couple of thoughts for future work (some you already know):
* We should think about removing the block entry and block context from the frontend
* We should remove finding the `<body>` function by name in b9run, and just execute function zero.
* We consider removing the ability to run a specific function from a script. It makes less sense now that symbol lookup is contextual. If we remove this functionality, it probably makes sense to remove the loop option too. After all, you can just call a function in a loop inside the script.
* A little hacky, but maybe we should compile the functions backwards, to maximize the impact of direct-call and pass-param.
